coreax/kernels/base.py

```diff
@@ -0,0 +1,503 @@
+``x`` and ``y``, and returns a value that decreases as ``x`` and ``y`` move further away
+in space from each other. Note that *further* here may account for cyclic behaviour in
+the data, for example.
+
+In this library, we often use kernels as a smoothing tool: given a dataset of distinct
+points, we can reconstruct the underlying data generating distribution through smoothing
+of the data with kernels.
+
+Some kernels are parameterizable and may represent other well known kernels when given
+appropriate parameter values. For example, the
+:class:`~coreax.kernels.SquaredExponentialKernel`,
+
+.. math::
+
+    k(x,y) = \text{output_scale} * \exp (-||x-y||^2/2 * \text{length_scale}^2),
+
+which if parameterized with an ``output_scale`` of
+:math:`\frac{1}{\sqrt{2\pi} \,*\, \text{length_scale}}`, yields the well known
+Gaussian kernel.
+
+A :class:`~coreax.kernels.ScalarValuedKernel` must implement a
+:meth:`~coreax.kernels.ScalarValuedKernel.compute_elementwise` method, which returns the
+floating point value after evaluating the kernel on two floats, ``x`` and ``y``.
+Additional methods, such as :meth:`ScalarValuedKernel.grad_x_elementwise`, can
+optionally be overridden to improve performance. The canonical example is when a
+suitable closed-form representation of a higher-order gradient can be used to avoid the
+expense of performing automatic differentiation.
+
+Such an example can be seen in :class:`coreax.kernels.SteinKernel`, where the analytic
+forms of :meth:`ScalarValuedKernel.divergence_x_grad_y` are significantly cheaper to
+compute than the automatic differentiated default.
+"""
+
+# Support class typing annotations inside itself
+from __future__ import annotations
+
+from abc import abstractmethod
+from typing import Union
+
+import equinox as eqx
+import jax
+import jax.numpy as jnp
+import jax.tree_util as jtu
+from jax import Array, grad, jacrev
+from jax.typing import ArrayLike
+from typing_extensions import override
+
+from coreax.data import Data, is_data
+from coreax.kernels.util import block_data_convert
+from coreax.util import pairwise, tree_leaves_repeat
+
+
+class ScalarValuedKernel(eqx.Module):
+    """Abstract base class for scalar-valued kernels."""
+
+    def __add__(self, addition: ScalarValuedKernel) -> AdditiveKernel:
+        """Overload `+` operator."""
+        if not isinstance(addition, ScalarValuedKernel):
+            raise ValueError("'addition' must be an instance of a 'ScalarValuedKernel'")
+        return AdditiveKernel(self, addition)
+
+    def __mul__(self, product: ScalarValuedKernel) -> ProductKernel:
+        """Overload `*` operator."""
+        if not isinstance(product, ScalarValuedKernel):
+            raise ValueError("'product' must be an instance of a 'ScalarValuedKernel'")
+        return ProductKernel(self, product)
+
+    def __pow__(self, power: int) -> ProductKernel:
+        """Overload `**` operator."""
+        min_power = 2
+        power = int(power)
```

Comment:
Note that doing `int(power)` just rounds `power` to the nearest integer - it won't throw any kind of error. To see this, change the test for invalid inputs to `__pow__` to instead to `kernel ** 2.5`, and note that no error is thrown (as the float is silently rounded to an int).

Reply:
I did realise this was happening but given that you cannot, in general, take fractional powers of kernels and [expect them to still be valid kernels](https://stats.stackexchange.com/questions/566822/in-machine-learning-for-a-kernel-function-k-is-sqrtk-also-a-valid-kernel-fun) I decided to round the float. I can see how this could be confusing and will make this error in the implementation of `PowerKernel`

Reply:
Ah, that's fair enough, I wasn't aware of that issue. Although just looking at the answer you linked, is the issue there that the exponent is less than 1, rather than that the exponent is fractional? I'm afraid I'm not familiar enough with kernels to answer that myself!

Reply:
The positive semi-definiteness of kernel functions is very related to the positive semi-definiteness of their resulting (kernel matrices)[https://en.wikipedia.org/wiki/Positive-definite_function]. So if we are considering fractional powers:

Given some kernel function $k:\mathcal{X}\times\mathcal{X}\to\mathbb{R}$ define

$$p(x,y) = k(x,y)^n$$

where

Reply:
Might be worth mentioning this in the `ScalarValuedKernel` or `ScalarValuedKernel.__pow__` docstring.